ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/Model.hs

```diff
@@ -446,18 +444,33 @@ addBlock cfg blk m = Model {
           (Proxy @(BlockProtocol blk))
           (selectView (configBlock cfg) . getHeader)
           (currentChain m)
-      . filter (extendsImmutableChain . fst)
-      $ candidates
+      $ consideredCandidates
 
-addBlocks :: LedgerSupportsProtocol blk
+    -- Sometimes blocks are added not in order and therefore we cannot just look
+    -- at the last added block and push it to the set of known valid blocks. The
+    -- easiest (and considering the lengths of the generated chains) way is to
+    -- just push all the valid blocks in the chain into the set.
+    --
+    -- Also, the chain selection algorithm in the SUT will test chains one by
+    -- one until it finds the one that should be selected, so prefixes of chains
+    -- will be tested for validity and blocks in those prefixes might be
+    -- included in the set of known valid blocks. Therefore, as we know which
+    -- chain was selected, and on every chain we know the valid blocks, we will
+    -- fold over the candidates up to (and including) the selected one.
+    valid' = foldl
+               (Chain.foldChain (\s b -> Set.insert (blockHash b) s))
+               (valid m)
+               (takeWhile (not . Chain.isPrefixOf newChain) (map fst consideredCandidates) ++ [newChain])
```

Comment:
Why do we need the `newChain` here? Shouldn't it be part of `consideredCandidates`

Reply:
Discussed with Javier: it took us >1 hour to convince ourselves that this correctly reflected the behaviour of the chain selection algorithm regarding the addition of valid blocks upon chain selection. So we should add some additional clarifications. Motivating example:

```
consideredCandidates

C0: vvvvxxxxxx
C1: vvvvvvx
C2: vvv

```

```
selectChain:

C1: vvvvvvx
C0: vvvvxxxxxx
C2: vvv

```

```
newChain: C1
```

Therefore:

```
new valid blocks: v's of C0 + v's of C1
```